### DISCUSSION TITLE
[Question] Inverse Kinematics Unitree G1 Robot

### DISCUSSION THREAD

**Discussion by francescotaioli** [isaac-sim/IsaacLab#1619] [Q&A]
Hi everyone,
I'm working on implementing Inverse Kinematics (IK) for the Unitree G1 Robot.

I began by referencing [docs](https://isaac-sim.github.io/IsaacLab/main/source/tutorials/05_controllers/run_diff_ik.html) and modified the script to focus on my desired joints. in particular, focusing on a simple test regarding the left part of the robot:
```
robot_entity_cfg = SceneEntityCfg(
            "robot",
            joint_names=[
                "left_shoulder_.*",
                "left_elbow_.*",
            ],
            body_names=["left_palm_link"],
        )
```
However, the results are not as expected. The issues I'm encountering are:

1. The right hand and elbows move when only the left side should be affected.
2. The robot is falling to the ground during the test.
3. The left hand is not reaching the target position

Ideally, I aim to achieve outcomes similar to those demonstrated [here](https://github.com/kevinzakka/mink?tab=readme-ov-file).

Does anyone know of additional documentation or scripts that could help? Any guidance would be greatly appreciated!

**Comment by RandomOakForest** [comment]
Thanks for posting this. With the latest release and the holidays we haven't been able to address this yet. I'll move it to a _discussion_ for the team to follow up upon.

**Comment by makolon** [comment]
Hi, @francescotaioli!
I’m also working on implementing inverse kinematics for the Unitree G1 robot with 23 degrees of freedom.

Firstly, regarding the 2 you mentioned, you can fix the Unitree G1 to the ground using the following script. By setting `fix_root_link=True` in the spawn function, you can prevent the G1 from falling to the ground:
```
        # Set Unitree G1 as robot
        self.scene.robot = G1_CFG.replace(
            prim_path="{ENV_REGEX_NS}/Robot"
        )
        self.scene.robot.init_state.pos = (2.75, 0.2, 0.74)
        self.scene.robot.init_state.rot = (1.0, 0.0, 0.0, 0.0)
        self.scene.robot.init_state.joint_pos = {
            ".*_hip_pitch_joint": -0.20,
            ".*_knee_joint": 0.42,
            ".*_ankle_pitch_joint": -0.23,
            ".*_elbow_pitch_joint": 0.0,
            ".*_elbow_roll_joint": 0.0,
            ".*_shoulder_roll_joint": 0.0,
            ".*_shoulder_pitch_joint": 0.0,
            ".*_shoulder_yaw_joint": 0.0,
            ".*_one_joint": 0.0,
            ".*_two_joint": 0.0,
            ".*_three_joint": 0.0,
            ".*_four_joint": 0.0,
            ".*_five_joint": 0.0,
            ".*_six_joint": 0.0,
        }
        # If you want to include leg joints, set kinematics to False
        if self.kinematic:
            self.scene.robot.spawn = sim_utils.UsdFileCfg(
                usd_path=f"{ISAACLAB_NUCLEUS_DIR}/Robots/Unitree/G1/g1.usd",
                activate_contact_sensors=True,
                rigid_props=sim_utils.RigidBodyPropertiesCfg(
                    disable_gravity=False,
                    retain_accelerations=False,
                    linear_damping=0.0,
                    angular_damping=0.0,
                    max_linear_velocity=1000.0,
                    max_angular_velocity=1000.0,
                    max_depenetration_velocity=1.0,
                ),
                articulation_props=sim_utils.ArticulationRootPropertiesCfg(
                    enabled_self_collisions=True,
                    solver_position_iteration_count=4,
                    solver_velocity_iteration_count=0,
                    fix_root_link=True,
                ),
            )
```

Additionally, I’ve noticed a minor issue when using the differential inverse kinematics controller with an absolute pose. Specifically, there is a small gap of approximately 1.1811 inches between the target and the actual end-effector pose.
```
        self.actions.left_arm_action = DifferentialInverseKinematicsActionCfg(
            asset_name="robot",
            joint_names=[
                "left_shoulder_pitch_joint",
                "left_shoulder_roll_joint",
                "left_shoulder_yaw_joint",
                "left_elbow_pitch_joint",
                "left_elbow_roll_joint",
            ],
            body_name="left_palm_link",
            controller=DifferentialIKControllerCfg(command_type="position", use_relative_mode=False, ik_method="dls"),
            body_offset=DifferentialInverseKinematicsActionCfg.OffsetCfg(pos=[0.0, 0.0, 0.0]),  # TODO: Check z-axis offset
        )
        self.actions.right_arm_action = DifferentialInverseKinematicsActionCfg(
            asset_name="robot",
            joint_names=[
                "right_shoulder_pitch_joint",
                "right_shoulder_roll_joint",
                "right_shoulder_yaw_joint",
                "right_elbow_pitch_joint",
                "right_elbow_roll_joint",
            ],
            body_name="right_palm_link",
            controller=DifferentialIKControllerCfg(command_type="position", use_relative_mode=False, ik_method="dls"),
            body_offset=DifferentialInverseKinematicsActionCfg.OffsetCfg(pos=[0.0, 0.0, 0.0]),  # TODO: Check z-axis offset
        )
```

**Reply by AndreSchakkal** [reply]
Hi, @makolon. I'm wondering which library are you using to implement the inverse kinematics for the Unitree G1 robot with 23 degrees of freedom?

**Reply by krishnagopal596** [reply]
Hey @AndreSchakkal, did you figure out which library is optimal for implementing Inverse Kinematics, I am working on the same robot but only for the Arms, 14 degrees of freedom, 7 for each arm. My implementation is not working as expected, can you give me any insights or the foundation for your approach? 
Thanks.

**Reply by AndreSchakkal** [reply]
Hi @krishnagopal596, in the end I did not pursue the idea of inverse kinematics... The closest I've done was Forward Kinematics by setting the joint angles of the robot in simulation and subsequently obtaining the desired body links locations.

**Reply by krishnagopal596** [reply]
Nice @AndreSchakkal, is it possible for you to share the github repo link of your implementation? My idea of using Forward Kinematics, was just to verify the joint angles by passing the end effector postions. Thank You.

**Comment by NocNoc31** [comment]
I am trying to find a way to solve IK for g1 using calculus, I hope everyone can help.

**Reply by krishnagopal596** [reply]
Hi @NocNoc31 , the last time I tried using Forward Kinematics, wasn't successfull, so it is still on hold.

**Comment by YuehChuan** [comment]
@NocNoc31  @francescotaioli 
unitree G1 29DOF did it in mujoco, 
Now my issue is the desire position and actual output has error, and how to control 7th hand pose.

https://github.com/YuehChuan/unitreeG1_ik/blob/main/g1_ik_solver.py
